### PR TITLE
feat(rules_check): validate the option examples for rules

### PR DESCRIPTION
**Pull request by cr7pt0gr4ph7**
## Summary

Extend the `rules_check` to:

(1) Ensure that configuration fragments are actually valid (and do not contain syntax or schema errors):

```rust
    /// ```json,options
    /// {
    ///     "options": { ... }
    /// }
    /// ```
```

(2) Validate rule examples that use a non-default configuration:

```rust
    /// ```json,options
    /// {
    ///     "options": { ... }
    /// }
    /// ```
    ///
    /// Invalid example for this configuration:
    ///
    /// ```js,expect_diagnostic,use_options
    /// import "invalid-import";
    /// ```
    ///
    /// Valid example for this configuration:
    ///
    /// ```json,use_options
    /// import "valid-import";
    /// ```
```

(3) Allow hiding lines from the output [like `rustdoc` does](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#hiding-portions-of-the-example) by prefixing them with `# `.

The corrresponding PR for [`biomejs/website`](https://github.com/biomejs/website) is biomejs/website#1442.

## Test Plan

- `cargo run -p rules_check`
- I manually inspected the documentation output generated for the website to ensure it is as intended

### CODE REVIEW

**Comment by cr7pt0gr4ph7** [comment]
Note: This PR would (partially) break the generated documentation until biomejs/website#1442 is merged. I would therefore recommend reviewing this PR here first, then reviewing & merging biomejs/website#1442, and then merging this PR (as outlined in [this issue comment](https://github.com/biomejs/website/pull/1442#issuecomment-2478452840)).

**Comment by cr7pt0gr4ph7** [comment]
As it turns out, I actually uncovered a bug where implementation & documentation differs for a rule when I un-ignored one of the examples for the `use_naming_convention` rule:

```
code-block.ts:2:20 lint/style/useNamingConvention ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! This interface name should be in PascalCase.
  
    1 │ declare module "myExternalModule" {
  > 2 │   export interface my_INTERFACE {}
      │                    ^^^^^^^^^^^^
    3 │ }
    4 │ 
  

Error: Analysis of 'style/useNamingConvention' on the following code block returned an unexpected diagnostic.
```

This is not actually related to the feature of validating options in the docs, so I have created issue #4545, referenced that issue in the documentation & ignored the offending doctest for now.

**Comment by cr7pt0gr4ph7** [comment]
> Few questions: (1) how would that work if there are multiple options for a rule, and we want to show multiple examples with different options? (2) is it possible to have multiple option blocks?

A `[language],use_options` block uses the options from the most recently preceding `[json_language],options` block [^1]. So the docs should look like this:

- Example options 1
- Valid or invalid code example for the given options 1
- Valid or invalid code example for the given options 1
- Valid or invalid code example for the given options 1
- Example options 2
- Valid or invalid code example for the given options 2
- etc. ...

A more complete example of a sequence of code blocks (within a single documentation comment):
| Block Options                      | Result                                                  |
| ---------------------------------- | ------------------------------------------------------- |
| `js`                               | (No Options)                                            |
| `js,use_options`                   | _Error: No active, valid options block_                 |
| `json,options`                     | = **Options 1**                                         |
| `js,use_options`                   | Use Options 1                                           |
| `js`                               | (No Options)                                            |
| `js,use_options`                   | Use Options 1                                           |
| `json,options`                     | **Options 2**                                           |
| `js,use_options`                   | Use Options 2                                           |
| `js,expect_diagnostic,use_options` | Use Options 2                                           |
| `json,expect_diagnostic,options`   | = **Example for Invalid Options**                       |
| `js,use_options`                   | ~Use Options 2~ _Error: No active, valid options block_ |

This works great for most examples, and encourages the style of `simple option example => (code example of that option in action)+`.

Further notes:
- I've only hit some issues for `useConsistentMemberAccessibility`, where it would probably be cleaner to show multiple diagnostics for a single example of a full class declaration, instead of 6 different but similar examples of members interleaved with their corresponding diagnostics output. Another idea would be to display the diagnostics output in a more compact manner, by making it collapsed-by-default, inlining it into the example code, ...

- The (not explicitly recommeded by [`biome_analyze/CONTRIBUTING.md`](https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md#document-the-rule) but still followed everywhere) documentation order of `Invalid Examples, Valid Examples, Options` does not work that well for  rules like `noRestrictedImports`, `noRestrictedTypes` etc. that _only_ make sense in combination with custom options.

[^1]: If no `options` block was encountered yet, or the last options block was a `[json_language],expect_diagnostic,options` (i.e. an invalid options example - this feature is implemented for the sake of completeness but currently not used), an error is reported on `use_options`.

File: crates/biome_analyze/CONTRIBUTING.md [hunk 1]
```diff
@@ -710,6 +710,8 @@ The documentation needs to adhere to the following rules:
 - When adding _invalid_ snippets in the `### Invalid` section, you must use the `expect_diagnostic` code block property. We use this property to generate a diagnostic and attach it to the snippet. A snippet **must emit only ONE diagnostic**.
 - When adding _valid_ snippets in the `### Valid` section, you can use one single snippet.
 - You can use the code block property `ignore` to tell the code generation script to **not generate a diagnostic for an invalid snippet**.
+- You can use the code block property `options` to tell the code generation script that this is a configuration options example.
+- You can use the code block property `use_options` to tell the code generation script to use the options from the most recent `options` block while linting.
```
[the remark this answers is not in the document]
Comment (cr7pt0gr4ph7) [reply crates/biome_analyze/CONTRIBUTING.md]: Done :+1:

**Review by arendjr — APPROVED** [review]

**Comment by ematipico** [comment]
Thank you, @cr7pt0gr4ph7, for the thorough explanation! Let's make sure to have such an explanation in the contribution guide because that's where the majority of our contributors learn how to contribute to our linter :)

**Review by ematipico — APPROVED** [review]

**Comment by cr7pt0gr4ph7** [comment]
> Let's make sure to have such an explanation in the contribution guide because that's where the majority of our contributors learn how to contribute to our linter :)

Already done :) See [the updated CONTRIBUTING.md](https://github.com/biomejs/biome/pull/4542/files#diff-5e4a05590c8f4b773454b32eb78a316f13a79d48453222f9ecee95c89dfd0ada). I also gave it a bit of additional polish (grammar, spelling, wording) and tried to make it better navigable for people new to Biome (like me). Is the updated contribution guide complete enough, or should I be even more explicit there?

**Comment by arendjr** [comment]
Thank you, @cr7pt0gr4ph7 ! This is a great improvement.

**Comment by Conaclos** [comment]
Thanks! This is a nice addition! :)

**Pull request merged**